# Data combining

## Preview raw data and combine raw data

In [3]:
import pandas as pd


In [6]:
preview=pd.read_csv('../../data/rawdata/T_DB1B_MARKET1.csv')
preview.head()

,ORIGIN_AIRPORT_ID,ORIGIN_AIRPORT_SEQ_ID,ORIGIN_CITY_MARKET_ID,DEST_AIRPORT_ID,DEST_AIRPORT_SEQ_ID,DEST_CITY_MARKET_ID,REPORTING_CARRIER,TICKET_CARRIER,OPERATING_CARRIER,BULK_FARE,PASSENGERS,MARKET_FARE,MARKET_DISTANCE,DISTANCE_GROUP,MARKET_MILES_FLOWN,NONSTOP_MILES,MKT_GEO_TYPE
0,10135,1013506,30135,10136,1013603,30136,OH,AA,99,0.0,1.0,895.41,1575.0,4,1575.0,1458.0,2
1,10135,1013506,30135,10140,1014005,30140,9E,DL,99,0.0,1.0,168.50,1961.0,4,1961.0,1738.0,2
2,10135,1013506,30135,10140,1014005,30140,9E,DL,99,0.0,1.0,201.84,1961.0,4,1961.0,1738.0,2
3,10135,1013506,30135,10140,1014005,30140,9E,DL,99,0.0,1.0,361.50,1961.0,4,1961.0,1738.0,2
4,10135,1013506,30135,10140,1014005,30140,9E,DL,99,0.0,1.0,843.00,1961.0,4,1961.0,1738.0,2


# Combining 1 year worth of data, q3 of 24 to q2 of 25

In [7]:


# List of CSV files to combine
csv_files = [
    "../../data/rawdata/T_DB1B_MARKET1.csv",
    "../../data/rawdata/T_DB1B_MARKET2.csv",
    "../../data/rawdata/T_DB1B_MARKET3.csv",
    "../../data/rawdata/T_DB1B_MARKET4.csv"
]

# Initialize an empty list to store dataframes
dfs = []

# Read and append each CSV file
for file in csv_files:
    print(f"Reading {file}...")
    df = pd.read_csv(file)
    print(f"  Shape: {df.shape}")
    dfs.append(df)

# Combine all dataframes
print("\nCombining all files...")
combined_df = pd.concat(dfs, ignore_index=True)

print(f"\nCombined dataset shape: {combined_df.shape}")
print(f"Memory usage: {combined_df.memory_usage(deep=True).sum() / 1024**2:.2f} MB")

# Save the combined dataset
output_path = "../../data/rawdata/T_DB1B_MARKET_COMBINED.csv"
print(f"\nSaving to {output_path}...")
combined_df.to_csv(output_path, index=False)

print("Done!")

Reading ../../data/rawdata/T_DB1B_MARKET1.csv...
  Shape: (7297028, 17)
Reading ../../data/rawdata/T_DB1B_MARKET2.csv...
  Shape: (8450420, 17)
Reading ../../data/rawdata/T_DB1B_MARKET3.csv...
  Shape: (8297869, 17)
Reading ../../data/rawdata/T_DB1B_MARKET4.csv...
  Shape: (8297869, 6)

Combining all files...

Combined dataset shape: (32343186, 17)
Memory usage: 7722.82 MB

Saving to ../../data/rawdata/T_DB1B_MARKET_COMBINED.csv...
Done!


In [8]:
combined_df=pd.read_csv('../../data/rawdata/T_DB1B_MARKET_COMBINED.csv')
combined_df.head()

C:\Users\Administrator\AppData\Local\Temp\ipykernel_48832\3461268546.py:1: DtypeWarning: Columns (6,7,8) have mixed types. Specify dtype option on import or set low_memory=False.
  combined_df=pd.read_csv('../../data/rawdata/T_DB1B_MARKET_COMBINED.csv')


,ORIGIN_AIRPORT_ID,ORIGIN_AIRPORT_SEQ_ID,ORIGIN_CITY_MARKET_ID,DEST_AIRPORT_ID,DEST_AIRPORT_SEQ_ID,DEST_CITY_MARKET_ID,REPORTING_CARRIER,TICKET_CARRIER,OPERATING_CARRIER,BULK_FARE,PASSENGERS,MARKET_FARE,MARKET_DISTANCE,DISTANCE_GROUP,MARKET_MILES_FLOWN,NONSTOP_MILES,MKT_GEO_TYPE
0,10135,1013506,30135,10136,1013603,30136,OH,AA,99,0.0,1.0,895.41,1575.0,4.0,1575.0,1458.0,2.0
1,10135,1013506,30135,10140,1014005,30140,9E,DL,99,0.0,1.0,168.50,1961.0,4.0,1961.0,1738.0,2.0
2,10135,1013506,30135,10140,1014005,30140,9E,DL,99,0.0,1.0,201.84,1961.0,4.0,1961.0,1738.0,2.0
3,10135,1013506,30135,10140,1014005,30140,9E,DL,99,0.0,1.0,361.50,1961.0,4.0,1961.0,1738.0,2.0
4,10135,1013506,30135,10140,1014005,30140,9E,DL,99,0.0,1.0,843.00,1961.0,4.0,1961.0,1738.0,2.0


## Data cleaning

In [4]:
# Data Loading
pd.set_option('display.max_columns', None)

df = pd.read_csv('../../data/rawdata/T_DB1B_MARKET_COMBINED.csv')   

# Data Overview
print("Data Loaded")
print(df.shape)
df.head()

C:\Users\Administrator\AppData\Local\Temp\ipykernel_17724\2164937600.py:4: DtypeWarning: Columns (6,7,8) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('../../data/rawdata/T_DB1B_MARKET_COMBINED.csv')


Data Loaded
(32343186, 17)


,ORIGIN_AIRPORT_ID,ORIGIN_AIRPORT_SEQ_ID,ORIGIN_CITY_MARKET_ID,DEST_AIRPORT_ID,DEST_AIRPORT_SEQ_ID,DEST_CITY_MARKET_ID,REPORTING_CARRIER,TICKET_CARRIER,OPERATING_CARRIER,BULK_FARE,PASSENGERS,MARKET_FARE,MARKET_DISTANCE,DISTANCE_GROUP,MARKET_MILES_FLOWN,NONSTOP_MILES,MKT_GEO_TYPE
0,10135,1013506,30135,10136,1013603,30136,OH,AA,99,0.0,1.0,895.41,1575.0,4.0,1575.0,1458.0,2.0
1,10135,1013506,30135,10140,1014005,30140,9E,DL,99,0.0,1.0,168.50,1961.0,4.0,1961.0,1738.0,2.0
2,10135,1013506,30135,10140,1014005,30140,9E,DL,99,0.0,1.0,201.84,1961.0,4.0,1961.0,1738.0,2.0
3,10135,1013506,30135,10140,1014005,30140,9E,DL,99,0.0,1.0,361.50,1961.0,4.0,1961.0,1738.0,2.0
4,10135,1013506,30135,10140,1014005,30140,9E,DL,99,0.0,1.0,843.00,1961.0,4.0,1961.0,1738.0,2.0


In [5]:
# Data Structure
df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32343186 entries, 0 to 32343185
Data columns (total 17 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   ORIGIN_AIRPORT_ID      int64  
 1   ORIGIN_AIRPORT_SEQ_ID  int64  
 2   ORIGIN_CITY_MARKET_ID  int64  
 3   DEST_AIRPORT_ID        int64  
 4   DEST_AIRPORT_SEQ_ID    int64  
 5   DEST_CITY_MARKET_ID    int64  
 6   REPORTING_CARRIER      object 
 7   TICKET_CARRIER         object 
 8   OPERATING_CARRIER      object 
 9   BULK_FARE              float64
 10  PASSENGERS             float64
 11  MARKET_FARE            float64
 12  MARKET_DISTANCE        float64
 13  DISTANCE_GROUP         float64
 14  MARKET_MILES_FLOWN     float64
 15  NONSTOP_MILES          float64
 16  MKT_GEO_TYPE           float64
dtypes: float64(8), int64(6), object(3)
memory usage: 4.1+ GB


,ORIGIN_AIRPORT_ID,ORIGIN_AIRPORT_SEQ_ID,ORIGIN_CITY_MARKET_ID,DEST_AIRPORT_ID,DEST_AIRPORT_SEQ_ID,DEST_CITY_MARKET_ID,BULK_FARE,PASSENGERS,MARKET_FARE,MARKET_DISTANCE,DISTANCE_GROUP,MARKET_MILES_FLOWN,NONSTOP_MILES,MKT_GEO_TYPE
count,3.234319e+07,3.234319e+07,3.234319e+07,3.234319e+07,3.234319e+07,3.234319e+07,2.404532e+07,2.404532e+07,2.404532e+07,2.404532e+07,2.404532e+07,2.404532e+07,2.404532e+07,2.404532e+07
mean,1.282563e+04,1.282567e+06,3.196636e+04,1.282417e+04,1.282421e+06,3.196489e+04,8.442392e-05,1.904901e+00,2.684043e+02,1.308387e+03,3.114319e+00,1.305384e+03,1.227264e+03,1.930577e+00
std,1.544638e+03,1.544636e+05,1.378709e+03,1.545144e+03,1.545143e+05,1.378273e+03,9.187862e-03,4.927751e+00,2.205487e+02,8.216434e+02,1.672517e+00,8.201554e+02,7.781802e+02,2.541717e-01
min,1.013500e+04,1.013506e+06,3.000700e+04,1.013500e+04,1.013506e+06,3.000700e+04,0.000000e+00,1.000000e+00,0.000000e+00,1.100000e+01,1.000000e+00,0.000000e+00,1.100000e+01,1.000000e+00
25%,1.129800e+04,1.129806e+06,3.072100e+04,1.129800e+04,1.129806e+06,3.072100e+04,0.000000e+00,1.000000e+00,1.475000e+02,7.170000e+02,2.000000e+00,7.130000e+02,6.590000e+02,2.000000e+00
50%,1.291500e+04,1.291503e+06,3.170300e+04,1.290200e+04,1.290205e+06,3.170300e+04,0.000000e+00,1.000000e+00,2.285000e+02,1.090000e+03,3.000000e+00,1.089000e+03,1.023000e+03,2.000000e+00
75%,1.410700e+04,1.410702e+06,3.310500e+04,1.410700e+04,1.410702e+06,3.310500e+04,0.000000e+00,1.000000e+00,3.370000e+02,1.759000e+03,4.000000e+00,1.754000e+03,1.660000e+03,2.000000e+00
max,1.686900e+04,1.686902e+06,3.642200e+04,1.686900e+04,1.686902e+06,3.642200e+04,1.000000e+00,1.032000e+03,3.873500e+04,1.134400e+04,2.300000e+01,1.134400e+04,9.411000e+03,2.000000e+00


In [6]:
print("\n===Missing Value Count===")
print(df.isna().sum())


===Missing Value Count===
ORIGIN_AIRPORT_ID              0
ORIGIN_AIRPORT_SEQ_ID          0
ORIGIN_CITY_MARKET_ID          0
DEST_AIRPORT_ID                0
DEST_AIRPORT_SEQ_ID            0
DEST_CITY_MARKET_ID            0
REPORTING_CARRIER        8297869
TICKET_CARRIER           8297869
OPERATING_CARRIER        8297869
BULK_FARE                8297869
PASSENGERS               8297869
MARKET_FARE              8297869
MARKET_DISTANCE          8297869
DISTANCE_GROUP           8297869
MARKET_MILES_FLOWN       8297869
NONSTOP_MILES            8297869
MKT_GEO_TYPE             8297869
dtype: int64


In [7]:
# Missing Value Example
df[df['REPORTING_CARRIER'].isna()].head()

,ORIGIN_AIRPORT_ID,ORIGIN_AIRPORT_SEQ_ID,ORIGIN_CITY_MARKET_ID,DEST_AIRPORT_ID,DEST_AIRPORT_SEQ_ID,DEST_CITY_MARKET_ID,REPORTING_CARRIER,TICKET_CARRIER,OPERATING_CARRIER,BULK_FARE,PASSENGERS,MARKET_FARE,MARKET_DISTANCE,DISTANCE_GROUP,MARKET_MILES_FLOWN,NONSTOP_MILES,MKT_GEO_TYPE
24045317,10135,1013506,30135,10136,1013603,30136,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
24045318,10135,1013506,30135,10140,1014005,30140,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
24045319,10135,1013506,30135,10140,1014005,30140,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
24045320,10135,1013506,30135,10140,1014005,30140,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
24045321,10135,1013506,30135,10140,1014005,30140,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [8]:
missing = df[df['REPORTING_CARRIER'].isna()]
print(missing['ORIGIN_AIRPORT_ID'].nunique(), "unique origin airports")
print(missing['ORIGIN_AIRPORT_ID'].value_counts().head(10))

447 unique origin airports
ORIGIN_AIRPORT_ID
11292    239079
12892    227918
13930    214651
12889    203318
10397    186739
14747    183678
13204    181860
11298    177599
14107    174452
14771    171039
Name: count, dtype: int64


realized a lot of US airports have missing info

In [9]:
origin_dest_pairs = missing.groupby(['ORIGIN_AIRPORT_ID', 'DEST_AIRPORT_ID']).size().reset_index(name='count')
origin_dest_pairs.sort_values(by='count', ascending=False, inplace=True)
origin_dest_pairs.head()

,ORIGIN_AIRPORT_ID,DEST_AIRPORT_ID,count
34463,12478,12892,7630
36394,12892,12478,7526
36328,12892,11618,7492
22027,11618,12892,7385
8759,10721,14771,7097


highest number of missing values for 
10721	Boston, MA: Logan International
11618	Newark, NJ: Newark Liberty International
12478	New York, NY: John F. Kennedy International
12892	Los Angeles, CA: Los Angeles International
14771	San Francisco, CA: San Francisco International


In [10]:
df_no_na = df.dropna()

In [11]:
df_no_na.info()
df_no_na.describe()
print("\n===Missing Value Count===")
print(df_no_na.isna().sum())

<class 'pandas.core.frame.DataFrame'>
Index: 24045317 entries, 0 to 24045316
Data columns (total 17 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   ORIGIN_AIRPORT_ID      int64  
 1   ORIGIN_AIRPORT_SEQ_ID  int64  
 2   ORIGIN_CITY_MARKET_ID  int64  
 3   DEST_AIRPORT_ID        int64  
 4   DEST_AIRPORT_SEQ_ID    int64  
 5   DEST_CITY_MARKET_ID    int64  
 6   REPORTING_CARRIER      object 
 7   TICKET_CARRIER         object 
 8   OPERATING_CARRIER      object 
 9   BULK_FARE              float64
 10  PASSENGERS             float64
 11  MARKET_FARE            float64
 12  MARKET_DISTANCE        float64
 13  DISTANCE_GROUP         float64
 14  MARKET_MILES_FLOWN     float64
 15  NONSTOP_MILES          float64
 16  MKT_GEO_TYPE           float64
dtypes: float64(8), int64(6), object(3)
memory usage: 3.2+ GB

===Missing Value Count===
ORIGIN_AIRPORT_ID        0
ORIGIN_AIRPORT_SEQ_ID    0
ORIGIN_CITY_MARKET_ID    0
DEST_AIRPORT_ID          0
DEST_AIRP

In [14]:
# Save the cleaned DataFrame to a new CSV file
df_no_na.to_csv('../../data/clean_data/T_DB1B_MARKET_CLEAN.csv', index=False)